PURE-2: Offline Updates (#3)
uptane/pures: pure2.md

* PURE: 2
* Title: Offline Updates
* Version: 1
* Last-Modified: 2021-09-16
* Authors: Jon Oster <jon.oster@toradex.com>, Phil Wise <phil@phil-wise.com>
* Status: Draft
* Content-Type: <text/markdown>
* Created: 2021-09-16

# PURE-2: Offline Updates (Draft)

## Abstract

> Firmware often needs to be updated offline, in situations where the full Uptane interactive flow is not possible for some reason--usually because internet connectivity is not available, or the device to be updated has not yet registered ECU keys with the Uptane backend. In those scenarios, it is still desirable to have some degree of validation and protection of the update process, and in particular it is desirable to ensure that these offline updates can be directed.

Reviewer: ```suggestion
Firmware often needs to be updated offline, in situations where the full Uptane interactive flow is not possible for some reason--usually because internet connectivity is not available, or the device to be updated has not yet registered ECU keys with the Uptane backend. In those scenarios, it is still desirable to have some degree of validation and protection for the update process to ensure that these offline updates can be directed.
```

Reviewer: I don't think the repetition of desirable is needed.

Author: Changed

> This PURE offers a method of creating a bundle of firmware images and metadata that can be put on durable storage and used to install updates in the absence of connectivity with the update server, but without turning that bundle into a permanent vector for rollback attacks. To accomplish this, we define a new role on the Director repository for signing targets metadata for offline installation, along with the procedures for verifying this metadata.

Reviewer 2: "targets metadata" -> "Targets metadata", here and throughout.

Author: Thanks for this (and the various other capitalization/style reminders), I'll fix them all in one commit later this week.

Reviewer: ```suggestion
This PURE offers a method of creating a bundle of firmware images and metadata that can be put on durable storage and used to install updates in the absence of connectivity with the update server without turning that bundle into a permanent vector for rollback attacks. To accomplish this, we define a new role on the Director repository for signing targets metadata for offline installation, along with the procedures for verifying this metadata.
```

Author: Fixed all of the above

## Motivation

> NOTE:   Although Uptane was initially designed for automotive applications, it has been adopted for use in other IoT domains. Therefore, instead of referring to "vehicles" as the target of updates, this PURE refers more generically to "devices".

Reviewer 2: I kind of wish this language was more generic everywhere in Uptane, although I imagine that the mental translation of "vehicles" to "devices" isn't so difficult.

There are a number of potential scenarios where an OEM may wish to validate and secure updates that are difficult to address with the full Uptane protocol. This is especially true in non-automotive uses of Uptane, such as in industrial automation systems or medical devices. Industrial automation systems may be on completely airgapped networks, never able to connect to update servers. Medical devices may be connected to the internet only very rarely, but still desire the advanced security benefits (in particular, selectively directed updates) that Uptane can provide over The Update Framework (TUF). Specific use cases for this PURE include:

-   Initial load-out of software on the factory floor, before all ECUs are present and before they are registered with the inventory database

-   Updating devices on airgapped networks where a full Uptane implementation may not be practical

-   Allowing service technicians to deliver updates at their discretion, but only from an OEM-directed set of possible updates

## Specification

### The offline-update metadata role 

> To solve the problems above, we add a new role to the Director repository, called the "offline-update" role. The new offline-update role is responsible for producing and signing metadata about sets of images that are valid for offline updates. This includes two types of metadata: offline-update snapshot metadata and offline-update targets metadata. These roles are analagous to snapshot and targets metadata on the image repository.

Reviewer 2: "Snapshot metadata", "Image repository", etc. Obviously not critical but good for consistency.

Reviewer: Since @pattivacek has called out the capitalizations in general, I am not addressing them here.

Author: fixed

Each offline-update targets metadata file lists a set of images eligible for installation together as an offline update. This metadata is designed to be loaded onto a portable storage medium, such as a USB drive or a diagnostics/flashing tool with attached storage, alongside an up-to-date version of all non-device-specific metadata from both Uptane repositories and binaries for all targets listed. Uptane clients can check for the presence of offline-update metadata, either by an automated process (e.g. periodically checking some well-known location on a filesystem or TFTP server) or via a manually-triggered diagnostic command. If offline-update metadata is found, the clients validate it, check it against the Image repository metadata, and then install the targets.

> When a devices receives online updates from the Director receive the latest version of the offline-update snapshot metadata as part of the normal update cycle.

Reviewer 3: This sentence does not make sense to me.

Author: Oops, let me fix that. I missed a word and a comma there; it should say "When a devices receives online updates from the Director, it receives the latest version of the offline-update snapshot metadata as part of the normal update cycle."

Reviewer 2: "When a device"

Author: 🤦

Author: fixed

### Offline-update role in root metadata

> To perform offline updates, the root metadata file in the Director repository SHALL include a new "offline-updates" role. This role MAY use the same signing key as the Targets role, but SHOULD use its own. The offline-update role's key(s) MAY be kept online or offline, depending on the security requirements of the implementor.

Reviewer 2: "its own" is a bit unclear, I'd specify something more like "a separate key" or "an independent key".

Reviewer 2: It also sounds to me like there are two roles here: Offline-Snapshot and Offline-Targets. Do they use the same key?

Author: This is my mistake; an earlier internal version had a singular role that produced/signed both types of metadata, because I couldn't see any security benefit to separating the roles. However, after doing some work on the implementation, it became clear that it makes it a lot easier on implementors if we maintain the 1:1 mapping between roles and metadata schema, so I changed so that there would be two roles. Evidently I didn't successfully change it everywhere.

Author: fixed

### Offline-update snapshot metadata

The offline-update snapshot metadata lists version numbers and filenames of all offline-update targets metadata files.

> For each offline-update targets metadata file on the repository, the offline-update snapshot metadata SHALL contain the following information:

Reviewer 2: I don't understand: if delegations aren't supported here, how will there ever be more than one Offline-Update Targets metadata files?

Reviewer 2: Ah never mind, I misread the bit below.

1.  The filename and version number of the offline-update targets metadata file.

### Offline-update targets metadata

Each offline-update targets metadata file on a repository contains information about a set of images eligible to be installed on devices. Like other Director targets metadata files, offline-update targets metadata files SHALL NOT contain delegations.

> When a device is validating offline-update targets metadata, it will only execute an update if it can unambiguously determine which images to install. For this reason, the metadata SHOULD include a mechanism to permit disambiguation in multi-ECU systems. See the deployment considerations section for more detail on this recommendation.

Reviewer: ```suggestion
When a device is validating offline-update targets metadata, it will only execute an update if it can unambiguously determine which images to install. For this reason, the metadata SHOULD include a mechanism to permit disambiguation in multi-ECU systems. See the Deployment Best Practices document for more detail on this recommendation.
```

Reviewer: I am assuming we will add a link to the appropriate section when this is finalized.

Author: I'm referring to the deployment considerations section of PURE-2 here, not the overall Deployment Best Practices document.

### Addendum to Director repository snapshot metadata requirements

> Offline-update targets metadata files SHOULD NOT be listed in the main snapshots metadata file of the Director repository, if implemented. Because the offline-update role signs its own snapshot metadata, listing those files in the main snapshots metadata file is redundant, and increases the size of a file that is downloaded very frequently.

Reviewer: ```suggestion
If implemented, offline-update targets metadata files SHOULD NOT be listed in the main snapshots metadata file of the Director repository. Because the offline-update role signs its own snapshot metadata, listing those files in the main snapshots metadata file is redundant, and increases the size of a file that is downloaded very frequently.
```

Author: I'm not a fan of this change, because it implies that "if implemented" is grammatically attached to "the Director repository". Although that reading is clearly nonsense to an experienced reader, I think my phrasing has less ambiguity.

### Required files for offline updates

> The following elements SHALL be present on any offline-update storage location (in a "well-known" location configured in advance on the on the Primary):

Reviewer 2: ```suggestion
The following elements SHALL be present on any offline-update storage location (in a "well-known" location configured in advance on the Primary):
```

-   Image repository metadata:

    -   All Root metadata files from the Image repository, from version 1 to the most-recent version at the time the offline update is generated

    -   The latest version of every Targets metadata file, including delegations, at the time the offline update is generated

    -   The latest version of the Snapshot metadata file at the time the offline update was generated

-   Director repository metadata:

    -   All Root metadata files from the Director repository, from version 1 to the most-recent version at the time the offline update is generated

    -   The offline-update Snapshot metadata file

>     -   Exactly one offline-update targets metadata file

Reviewer 3: What happens if there is  more than one?

Author: The client would take the first one listed in the snapshot metadata:

> 4. Traverse the list of offline-update targets metadata files in the offline-update Snapshot metadata until a matching filename is found in the well-known location.

I do think it might be useful to be able to include multiple options in one usb stick and have a user-controlled selection mechanism, but it would have complicated the spec to include user interaction/selection as part of the process. My thinking here was that an implementor could accommodate this use case just by making several different PURE-2 directories, and having the user select the one they want before the process starts, e.g.:

```
.well-known
├── option-1
│   ├── images
│   │   ├── firmware-acme-1.0.2.bin
│   │   └── firmware-bravo-3.1.1.bin
│   └── metadata
│       ├── director
│       │   ├── EMEA-standard.json
│       │   ├── offline-update-snapshot.json
│       │   ├── 1.root.json
│       │   ├── 2.root.json
│       │   ├── 3.root.json
│       │   ├── 4.root.json
│       │   └── 5.root.json
│       └── image-repo
│           ├── delegation-1.json
│           ├── delegation-2.json
│           ├── 1.root.json
│           ├── 2.root.json
│           ├── 3.root.json
│           ├── snapshot.json
│           └── targets.json
└── option-2
    ├── images
    │   ├── firmware-charlie-1.0.2.bin
    │   └── firmware-delta-3.1.1.bin
    └── metadata
        ├── director
        │   ├── APAC-standard.json
        │   ├── offline-update-snapshot.json
        │   ├── 1.root.json
        │   ├── 2.root.json
        │   ├── 3.root.json
        │   ├── 4.root.json
        │   └── 5.root.json
        └── image-repo
            ├── delegation-1.json
            ├── delegation-2.json
            ├── 1.root.json
            ├── 2.root.json
            ├── 3.root.json
            ├── snapshot.json
            └── targets.json
```

-   All target images referenced in the offline-update targets metadata file

### Metadata verification procedures

The procedures for verifying offline-update metadata are similar to the procedures for verifying normal director metadata. However, there are some key differences. The procedures for sending images from the primary to secondaries, as well as for verifying and installing the images once the metadata has been verified, remain the same.

> Note: wherever the "Uptane Standard" or "the Standard" is referenced, it refers to version 1.2.0 of the Standard, available at https://uptane.github.io/uptane-standard/1.2.0/uptane-standard.html.

Reviewer 2: Does this mean we will need to update this once v2 is released? How do we handle that going forward?

Author: Good question. This note might not even be necessary.

Reviewer 4: If v2 comes out before this PURE is accepted, we can update this here. Otherwise once this is accepted and included in the Standard I think it's fine to let the PURE reference an old version of the Standard.

Reviewer 2: I think it's not a bad idea to reference the specific version the PURE was written in mind for. That way if there is any ambiguity, it's at least clear what version was intended, and if it needs to be updated later to resolve compatibility issues, it's easier to know what it needs to compare with.

The primary still SHALL verify all target images listed in the Director offline-update metadata, while secondaries MAY verify only the images they will install.

#### Download and verify metadata from the well-known location

The Primary SHALL download metadata for all offline-update targets and perform a full verification on it. Full verification of metadata means that the ECU checks that the Targets metadata about images from the Director repository matches the Targets metadata about the same images from the Image repository. This provides resilience to a key compromise in the system.

> Because of the nature of offline updates, sometimes the offline-update image will contain newer metadata than the primary has, and sometimes the converse will be true. Most of the differences or new steps in the verification process are to deal with this reality; the rest of the differences are steps that were removed because they aren't applicable, mostly relating to the ECU version reports and vehicle version manifest.

Reviewer 2: ```suggestion
Because of the nature of offline updates, sometimes the offline-update image will contain newer metadata than the primary has, and sometimes the converse will be true. Most of the differences or new steps in the verification process are to deal with this reality; the rest of the differences are steps that were removed because they aren't applicable, in particular relating to the ECU version reports and vehicle version manifest.
```
Something about "mostly" sounds too hand-wavy to me.

> Because of the nature of offline updates, sometimes the offline-update image will contain newer metadata than the primary has, and sometimes the converse will be true. Most of the differences or new steps in the verification process are to deal with this reality; the rest of the differences are steps that were removed because they aren't applicable, in particular relating to the ECU version reports and vehicle version manifest.

Reviewer: ```suggestion
Because of the nature of offline updates, sometimes the offline-update image will contain newer metadata than the primary, and sometimes the converse will be true. Most of the changed or added steps in the verification process were made to deal with this reality; the rest of the differences are steps that were removed because they are no longer applicable, in particular those relating to the ECU version reports and vehicle version manifest.
```

Author: Applied

In order to perform offline-update verification, an ECU SHALL perform the following steps.

1.  Load and verify the current time or the most recent securely attested time.

2.  Check if updated Director root metadata is available in the well-known location, following the procedure listed in Section 5.4.4.3 of the Standard, loading the root metadata file(s) from the well-known location instead of attempting to download them from the repository.

3.  Update offline-update Snapshot metadata from the well-known location:

>     1.  Load the offline-update Snapshot metadata file from the well-known location. If the version number of the existing offline-update Snapshot metadata file is equal to or greater than the one found in the well-known location, continue to step 4.

Reviewer 2: "existing"? You mean previously stored? The phrasing here is ambiguous because we're talking about two local storage locations, so we should be really clear about what to get from where.

Author: Fixed

4.  Traverse the list of offline-update targets metadata files in the offline-update Snapshot metadata until a matching filename is found in the well-known location. If no matching file is found, abort the update cycle and log the error. If one is found, perform the following verification procedure:

    1.  Check that the version number of this Targets metadata file matches the version number listed in the latest Snapshot metadata. If the version number does not match, discard it, abort the update cycle, and log the failure.

    2.  Check that the Targets metadata has been signed by the threshold of unique keys specified for the offline-updates Targets role in the latest Root metadata file. If not, discard it, abort the update cycle, and log the failure. (Checks for an arbitrary software attack.)

    3.  Check that the current (or latest securely attested) time is lower than the expiration timestamp in this Targets metadata file. If not, abort the update cycle and log the failure. (Checks for a freeze attack.)

>     4.  Check that there is at most one target in the metadata that can be installed on this ECU. The exact method for this will vary between implementations. For example, if each target specified a hardwareId, make sure that there is only one target listed for this ECU's hardwareId. If there is more than one target that could be installed on this ECU, abort the update cycle and log the failure (if this is the primary ECU) or report the failure to the primary ECU (if this is a secondary ECU).

Reviewer 2: ```suggestion
    4.  Check that there is at most one target in the metadata that can be installed on this ECU. The exact method for this will vary between implementations. For example, if each target specified a hardware identifier, make sure that there is only one target listed for this ECU's hardware identifier. If there is more than one target that could be installed on this ECU, abort the update cycle and log the failure (if this is the primary ECU) or report the failure to the primary ECU (if this is a secondary ECU).
```

> 5.  If the offline-update Targets metadata indicates that there is no new target for this ECU (either because there is no target valid for install on this ECU, or because the target is already installed), the verification process MAY be stopped here and considered complete. Otherwise, continue with the rest of the steps.

Reviewer: ```suggestion
5.  If the offline-update Targets metadata indicates that there is no new target for this ECU (either because there is no target valid for installation on this ECU, or because the target is already installed), the verification process MAY be stopped here and considered complete. Otherwise, continue with the rest of the steps.
```

Author: applied

3.  Check that the version number listed by the previous Snapshot metadata file for each Targets metadata file is less than or equal to its version number in this Snapshot metadata file. If this condition is not met, discard the new Snapshot metadata file, abort the update cycle, and log the failure. (Checks for a rollback attack.)

### Addendum to Uptane Standard Build-time prerequisite requirements for ECUs

For an ECU to be capable of receiving PURE-2-secured offline updates, it MUST have the following data provisioned at the time it is manufactured or installed in the vehicle:

1. A sufficiently recent copy of required Uptane metadata at the time of manufacture or install. This is necessary for the ECU to authenticate that the remote repository is legitimate when it first downloads metadata in the field. See [Uptane Deployment Best Practices](https://uptane.github.io/deployment-considerations/index.html) for more information. 
2. The current time, or a secure attestation of a sufficiently recent time.

The ECU does not require an ECU key in order to be capable of receiving PURE-2-secured offline updates.

### Sample directory structure

> Note that, although the offline-update snapshot metadata file lists two possible offline updates (EMEA-standard.json and EMEA-premium.json, see sample metadata below), a single source of offline-update metadata SHOULD always only include one of them--otherwise, the device wouldn't know which one to install. An implementor MAY choose to allow multiple offline-update targets metadata files in the same location, however: for example, if the Uptane client on the primary has the capability to display different options and allow the technician or end-user to select the one they wish to install.

Reviewer 2: Maybe this note should be closer to the Snapshot metadata example file? Here it's a bit confusing. I also think the last sentence is structured in a fairly convoluted manner.

Reviewer: ```suggestion
Note that, although the offline-update snapshot metadata file lists two possible offline updates (EMEA-standard.json and EMEA-premium.json, see sample metadata below), a single source of offline-update metadata SHOULD always only include one of them--otherwise, the device wouldn't know which one to install. However, an implementor MAY choose to allow multiple offline-update targets metadata files in the same location. For example, if the Uptane client on the primary has the capability, it could display different options and allow the technician or end-user to select the one they wish to install.
```

Reviewer: I agree with @pattivacek. The last sentence was rather hard to follow. I have suggested a fix.

Author: Addressed both

```
.well-known
├── images
│   ├── firmware-acme-1.0.2.bin
│   └── firmware-bravo-3.1.1.bin
└── metadata
    ├── director
    │   ├── EMEA-standard.json
    │   ├── offline-update-snapshot.json
    │   ├── 1.root.json
    │   ├── 2.root.json
    │   ├── 3.root.json
    │   ├── 4.root.json
    │   └── 5.root.json
    └── image-repo
        ├── delegation-1.json
        ├── delegation-2.json
        ├── 1.root.json
        ├── 2.root.json
        ├── 3.root.json
        ├── snapshot.json
        └── targets.json
```

### Sample metadata files

> These three examples are given because they're the only files that are unique to the offline update proposal. All the other metadata files in the directory structure are the same as existing Uptane metadata.

Reviewer 2: ```suggestion
These three examples are given because they are the only files that are unique to the offline update proposal. All the other metadata files in the directory structure are the same as existing Uptane metadata.
```

#### metadata/director/EMEA-standard.json

> This sample offline-update targets metadata file would direct a device to install firmware-acme-1.0.2.bin on the acme-flibberator-NBB2 ECU, and firmware-bravo-3.1.1.bin on the bravo-turboencabulator ECU, regardless of the specific ECU serial number of those devices. That installation directions is only valid until 2022-02-24T20:54:38Z, and furthermore is only valid if version 5 of offline-update-EMEA.json is the version that the device's latest offline-update-snapshot.json file indicates.

Reviewer 2: ```suggestion
This sample offline-update targets metadata file would direct a device to install firmware-acme-1.0.2.bin on the acme-flibberator-NBB2 ECU, and firmware-bravo-3.1.1.bin on the bravo-turboencabulator ECU, regardless of the specific ECU serial number of those devices. These installation directions are only valid until 2022-02-24T20:54:38Z, and furthermore are only valid if version 5 of offline-update-EMEA.json is the version that the device's latest offline-update-snapshot.json file indicates.
```

                    "hardwareIds":
                    [
                        "acme-flibberator-NBB2"
                    ],
                    "targetFormat": "BINARY",
                    "uri": null,
                    "createdAt": "2021-04-05T14:24:00Z",
                    "updatedAt": "2021-04-05T14:24:00Z",
                }
            },
            "firmware-bravo-3.1.1.bin":
            {
                "hashes":
                {
                    "sha256": "5614fa9539dd64735ac5afe9f81a6fafb24998f32ef0eaa62d7e88e422272d2b"
                },
                "length": 1011928,
                "custom":
                {
                    "name": "firmware-bravo",
                    "version": "3.1.1",
                    "hardwareIds":
                    [
                        "bravo-turboencabulator"
                    ],
                    "targetFormat": "BINARY",
                    "uri": null,
                    "createdAt": "2021-04-05T14:24:47Z",
                    "updatedAt": "2021-04-05T14:24:47Z",
                }
            },
        "version": 5
    }
}
```

#### metadata/director/offline-update-snapshot.json

> This sample offline-update snapshot metadata file indicates to the device that the latest version of offline-update-EMEA.json is 5, and the latest version of offline-update-APAC.json is 3 (and also indicates the hashes of those files, though this is not required by PURE-2). The device SHALL reject any other version of either of those files. It also SHALL reject any offline updates after 2021-11-21T15:38:10Z. Note that this doesn't mean that offline-update-EMEA.json version 5 and offline-update-APAC.json version 3 stop being valid on that date: if the device already has valid offline-update snapshot metadata with a higher version number than this (which might have been received from an OTA update or from another offline update), it can still use the versions of offline-update-EMEA.json and offline-update-APAC.json on this USB stick if the newer offline-update-snapshot.json indicates those are still the latest versions of those files.

Reviewer 2: The USB stick reference comes a bit out of nowhere. I'd suggest using consistent terminology about known locations or portable storage media.

Reviewer 2: Here (and above) you mention `offline-update-APAC.json`, but the actual example has `EMEA-premium.json`, which is also referred to above.

Reviewer: ```suggestion
This sample offline-update snapshot metadata file indicates to the device that the latest version of offline-update-EMEA.json is 5, and the latest version of offline-update-APAC.json is 3 (and also indicates the hashes of those files, though this is not required by PURE-2). The device SHALL reject any other version of either of those files. It also SHALL reject any offline updates after 2021-11-21T15:38:10Z. Note that this doesn't mean that offline-update-EMEA.json version 5 and offline-update-APAC.json version 3 stop being valid on that date. If the device already has valid offline-update snapshot metadata with a higher version number than this (which might have been received from an OTA update or from another offline update), it can still use the versions of offline-update-EMEA.json and offline-update-APAC.json on this USB stick if the newer offline-update-snapshot.json indicates those are still the latest versions of those files.
```

Author: I revamped this quite a bit, but I think it's a lot stronger now. I moved the text giving the example of how it could work to have a newer version of snapshot so that it's after the example file, and split it into several different sentences for clarity.

In general, we recommend being conservative and careful about the scenarios in which offline updates are accepted, because of the risk that an attacker could find a way to exploit the well-known location and install an undesirable (but not-yet-expired) offline update.

### Preparing an ECU for PURE-2

The prerequisites for PURE-2 updates are similar to the prerequisites for normal Uptane updates, except that the ECU does not need to have an ECU key yet. This follows from the fact that no ECU manifests or vehicle version reports are created during an offline update, and is useful in factory programming use cases, where ECUs may not yet have their private key material provisioned in before receiving a software update.

> ### Partial-verification ECUs and PURE-2

Reviewer 2: ```suggestion
### Partial verification ECUs and PURE-2
```

It is possible for a partial-verification ECU to perform PURE-2 offline updates. Like partial verification for normal Uptane updates, a partial-verification ECU MAY verify only the offline update targets metadata file it receives from the primary.

### Metadata expiry times

> The Offline Updates role signs two different types of metadata: offline-update targets metadata files, which contain the targets to be installed, and the the offline-update snapshot metadata, which lists the currently-valid offline-update targets metadata files. For an offline update to be installed, both the offline update snapshot metadata file and the targets metadata file being installed MUST be within their validity period.

Reviewer: ```suggestion
The Offline Updates role signs two different types of metadata: offline-update targets metadata files, which contain the targets to be installed, and the offline-update snapshot metadata, which lists the currently-valid offline-update targets metadata files. For an offline update to be installed, both the offline-update snapshot metadata file and the targets metadata file being installed MUST be within their validity period.
```

Author: fixed

> Therefore, choosing the expiry times for these files is an important security consideration, and an important usability consideration. Making the right choice will depend heavily on the implementation and what purpose offline updates are serving. For example, if offline updates are designed for 

Reviewer 2: Eh?

Author: Whoops, looks like the rest of this paragraph got lost in the ether somewhere. Will finish it up shortly.

Author: fixed...

> Will finish it up shortly.

....on Nov. 23. That still counts as "shortly" on a standards-approval timescale, right? :-/

## Security Analysis

This proposal does not detract from existing security guarantees because it does not modify the normal Uptane verification procedures for online updates.

However, offline updates under PURE-2 do offer somewhat different security guarantees than online updates.

### Alternatives considered

> In developing this PURE, the possibility of simply using Image repository metadata for offline updates was considered. The main argument in favour of taking this approach is that the Image repo's intended purpose is to attest _validity_ of firmware to be installed, and that in cases where offline updates are permitted/desired, the person applying the update will have the knowledge and ability to determine which specific targets to install. (This would effectively mean that offline updates would just use The Update Framework.)

Reviewer 2: ```suggestion
In developing this PURE, the possibility of simply using Image repository metadata for offline updates was considered. The main argument in favour of taking this approach is that the Image repo's intended purpose is to attest _validity_ of firmware to be installed, and that in cases where offline updates are permitted/desired, the person directing the update will have the knowledge and ability to determine which specific targets to install. (This would effectively mean that offline updates would just use The Update Framework.)
```
I think of applying as a process on the ECU, akin to flashing. Here we are talking about the person who's deciding what the next update should be. "Directing" may still not be the best word choice.

Author: Hmm...I don't really like "directing" here either because of the parallel to the director repository. I don't actually think there's a conflict, though: although applying/flashing/whatever is a process on the ECU, there's still a human doing it (in this case). In the same way that we say "I copied a file" instead of "I directed my operating system to copy a file", I think it's contextually very clear what's meant here.

Reviewer 2: Yeah, it's probably fine. I can't think of a better word, anyway.

Reviewer: Being a bit nit-picky here, but we did agree in the style sheet to U.S. spellings.
```suggestion
In developing this PURE, the possibility of simply using Image repository metadata for offline updates was considered. The main argument in favor of taking this approach is that the Image repository's intended purpose is to attest _validity_ of firmware to be installed, and that in cases where offline updates are permitted/desired, the person applying the update will have the knowledge and ability to determine which specific targets to install. (This would effectively mean that offline updates would just use The Update Framework.)
```

Author: applied

This approach is certainly possible, and indeed can be implemented without an extension to Uptane. It also represents an improvement over the status quo of many real-world implementations. However, PURE-2 introduces important additional protections.

> * A TUF-only approach does offer any protection against rollback attacks, unless the Image repository is regularly pruned of older images. Because a key element of Uptane is the signed Vehicle Version Manifest, Uptane installations often keep older, potentially-exploitable firmware images in their image repositories as long as there are still some extant devices that are still running the old image. Thus, a more robust mechanism is desirable to protect against rollbacks.

Reviewer: A TUF-only approach does offer or does not offer? If it does, the word "any" needs to be removed.

Author: Should say "does not", thanks. Fixed.

* With PURE-2, it is possible to fine-tune the expiry times of offline update metadata. A TUF-only approach means that a downloaded offline update only remains valid as long as the Image repository Targets metadata remains valid, and many Uptane implementations choose short expiry times. Expressing the expiry of a bundle of images to install independently while maintaining the ability to get newer Targets metadata out of band is a significant ergonomic improvement without necessitating the major security concession of choosing longer Targets metadata expiry times.
* Directing/specifying a specific bundle of images to be installed is a very important safety, security, and regulatory concern in multi-ECU systems; using a TUF-only approach relies on the person installing the updates to make correct compatibility choices.

> ### Understanding of device fleets' possible install base

Reviewer: Maybe it's just my ignorance, but this subhead makes no sense to me. What does this have to do with fleets?

Author: You're right, the heading isn't very good. I've changed it to "Restricting side-channel updates", which is more central to the point of the section. Having unknown images in the fleet's install base is a possible consequence of having updates that were installed outside of uptane, and PURE-2 reduces the need for those side channels.

Any offline update scheme must deal with the fact that the Uptane server will not have perfectly up-to-date information on the firmware versions of devices in the field. Currently, Uptane implementors are left without guidance on how offline updates should be implemented, and the result of this gap is a patchwork of security measures (or non-measures). It is still relatively common for vehicle ECUs to have a flashing mode secured by nothing more than the inconvenience of physically accessing a particular jumper, and signature checking when flashing firmware via diagnostic tool is still quite uncommon. The practical upshot of this is that operators of Uptane systems must consider the possibility that any arbitrary software could be installed out-of-band.

> PURE-2 limits the scope of what may have been installed on a device. If the Uptane client on the device supports PURE-2, it is much more feasible to completely lock down alternate methods of flashing: removing JTAG headers, removing alternate low-level flashing routines from firmware, etc. Because the Offline Update role signs all metadata authorizing updates outside of the normal online process, the scope of possibility of what may be installed in the field can be greatly reduced. Furthermore, because the offline updates are being facilitated by the same client and within the same framework as the OTA updates, it becomes much easier to maintain an accurate and complete history of software updates on the device.

Reviewer: ```suggestion
PURE-2 limits the scope of what may have been installed on a device. If the Uptane client on the device supports PURE-2, it is much more feasible to completely lock down alternate methods of flashing by removing JTAG headers, and/or alternate low-level flashing routines from firmware, etc. Because the Offline Update role signs all metadata authorizing updates outside of the normal online process, the scope of what may be installed in the field can be greatly reduced. Furthermore, because the offline updates are being facilitated by the same client and within the same framework as the OTA updates, it becomes much easier to maintain an accurate and complete history of software updates on the device.
```

Author: Applied a slightly different version of this suggestion

### Revocation of metadata

Like all Uptane metadata, offline-update metadata can be revoked both implicitly (by allowing it to expire) and explicitly (by issuing new metadata superseding and invalidating the old). Explicit revocations will reach devices when they check for an online update, because they always download the latest version of the offline-update snapshot metadata during an update check. However, if the device does not check for updates for a long time, it will not be aware of newer metadata that includes the revocation.

### Validation and updating of metadata

> An offline-update image also includes metadata for other Uptane roles:

Reviewer 2: The image itself contains metadata? Like, hardcoded updates to the offline-update metadata, or what? I'm confused.

Author: Yeah, we have some terminology stuff to take care of here. "Image" here refers to the complete bundle of stuff you stick in the well-known location for the update client to find, not to some particular binary you're installing. That's probably not ideal. Maybe "bundle"?

Reviewer 2: "Bundle" sounds pretty good, as long as that term is defined somewhere.

Reviewer: Just an FYI @tkfu and @pattivacek  "Bundle" is defined in the Glossary found in the Deployment pages

Author: I looked through the document, and there was only one other place where I'd used "image" like this, so I elected to replace it with the word "bundle", and define it parenthetically in-line in both places.

* The root role from both Director and Image
* The snapshot role from Image
* All targets roles from Image

> During an offline update, the device will only accept the metadata being offered in the offline update if it is newer than the metadata the device already has. Note that this does not imply that an offline update will fail if the device has a newer version of some metadata than the version on offer. For an offline-update targets metadata file to be accepted by the device, the following conditions MUST be met:

Reviewer: ```suggestion
During an offline update, the device will only accept the metadata being offered if it is newer than the metadata the device already has. Note that this does not imply that an offline update will fail if the device has a newer version of some metadata than the version on offer. For an offline-update targets metadata file to be accepted by the device, the following conditions MUST be met:
```

Author: applied

* The file itself has not expired
* The latest valid offline-update Snapshot metadata file the device possesses has not expired
* The file is signed by the threshold of unique keys listed in the latest Director root metadata the device possesses for the Offline-update Targets role
* The file is listed in the latest offline-update Snapshot metadata file the device possesses, and matches the version number listed in that Snapshot metadata
* All targets that appear in the file are also present in the latest version of Image repo Targets metadata the device possesses

> As a consequence, an individual offline-update targets metadata file will be accepted until it's invalidated by one of the following methods:

Reviewer 2: ```suggestion
As a consequence, an individual offline-update targets metadata file will be accepted until it is invalidated by one of the following methods:
```

* It expires
* Director's offline-update Snapshot metadata expires
* A new version of offline-update Snapshot metadata is issued with the file removed or bumped to a later version, AND that new version reaches the device by some means (either online or offline update)
* The offline-update targets signing keys are changed in Director root metadata AND the new metadata reaches the device by some means (either online or offline update)

> * An individual target listed in the file is removed from the Image repo Targets metadata AND the new metadata reaches the device by some means (either online or offline update)

Reviewer: I think we should always spell out the word repository. The Standard should be formal in approach and so I don't think "repo" is appropriate. I corrected one or two already, but thought I should just flag this as something to fix throughout.

Author: done
